app/models/host/managed.rb

```diff
@@ -871,6 +873,37 @@ def validate_dns_name_uniqueness
     end
   end
 
+  def update_external_ipam
+    # Return immediately if there are any model errors from upstream validation
+    return unless self.errors.full_messages.empty?
+
+    if plugin_present?('foreman_ipam')
+      self.interfaces.each do |interface|
+        ip = interface['ip']
+        is_new = interface['id'].nil?
+        subnet = interface['subnet_id'].nil? ? nil : Subnet.find(interface['subnet_id'].to_i)
+
+        if subnet&.ipam&.downcase == "external ipam"
+          old_ip = Nic::Base.find(interface['id'].to_i)['ip'] unless is_new
+          ip_changed = old_ip != ip unless is_new
+
+          if is_new || (!is_new && ip_changed)
+            response = subnet.external_ipam_proxy.ip_exists(ip, subnet)
+            ip_used = response['exists'] == true
+
+            if !ip_used
+              subnet.external_ipam_proxy.add_ip_to_subnet(ip, subnet)
+              subnet.external_ipam_proxy.delete_ip_from_subnet(old_ip, subnet) if ip_changed
+            else
+              interface.errors.add(:ip, :taken)
+              self.errors.add :interfaces, _('Some interfaces are invalid')
+            end
+          end
+        end
+      end
+    end
+  end
+
```

Comment:
So we have an orchestration "framework" in Foreman that you should use. Search for `orchestration/dhcp.rb` for an example, you should create new file `ipam.rb` there and queue create, update and delete methods the same way as dhcp does. It does support transaction compensation - when something goes wrong it performs a rollback.